Added docs deployment step in GHA

vfdev-5 opened the pull request:
Description:
- Added GHA "docs deploy" job based on "build-docs" job
- Using an "action" to commit and push to the branch.
  - use `git reset --soft <fixed-commit>` to squash history
  - History on gh-pages: https://github.com/vfdev-5/functorch/commits/gh-pages
- Runned GHA that pushed to gh-pages on my fork:
  - https://github.com/vfdev-5/functorch/actions/runs/1888876266


Fixes #500

zou3519 commented on .github/workflows/docs.yml:
```diff
@@ -35,5 +35,39 @@ jobs:
+          # Manipulations to squash the history:
+          # 1) reset to a fixed commit
+          # 2) use github-actions-x/commit to commit everything as a single commit
+          git reset --soft 218f32faf8b8cc5236bc49ea77ef7d222e21e5d6
```
What is this commit?

vfdev-5 replied:
https://github.com/pytorch/functorch/commit/218f32faf8b8cc5236bc49ea77ef7d222e21e5d6

zou3519 replied:
If nothing changes on the docs between to commits to pytorch/pytorch, is the commit essentially empty or is there something in it?

If it is essentially empty, then I'm thinking we should comment this out for now and then turn it on later -- I want to be able to manually submit changes to the ghpages branch for the next couple of weeks.

vfdev-5 replied:
This actions is triggered on every commit regardless python docstring or docs/source is unchanged (to make this action aware of specific files changes we can specify that explicitly.)
I think it could be difficult to check at runtime if commit changes or not the docs.
We may opt to another solution using another sub-action: https://github.com/devops-infra/action-commit-push which should allow `--amend` commit option. So, it means when this actions is run and rebuild the docs it would commit ammend the results on gh-pages to the last commit. What do you think ?

zou3519 replied:
Hmm, my question is, is the commit "empty" if nothing changes? Does the docs build generate different documentation based on the time of day or does it always generate the same thing if nothing has changed?

I'm trying to evaluate how large of a problem the history will become -- a lot of commits that have very little changes in them (due to us not updating the website) means we don't need to worry about having a lot of commits.

zou3519 reviewed (approved):
This LGTM. I want to wait until the website is a little more stable before merging this

zou3519 commented on .github/workflows/docs.yml:
```diff
@@ -35,5 +35,39 @@ jobs:
+      - name: Copy built docs to nightly
+        run: |
+          cp -R /tmp/docs/* nightly/
+          # Manipulations to squash the history:
+          # 1) reset to a fixed commit
+          # 2) use github-actions-x/commit to commit everything as a single commit
+          git reset --soft 218f32faf8b8cc5236bc49ea77ef7d222e21e5d6
```
I thought about it some more, and I think it would be better if we didn't reset the history for now (but turned it on later). The reasoning is:
1. We might do manual commits to the docs from time-to-time
2. We don't actually produce that much docs, so it'll take a while (a couple of months?) before we actually run into memory problems

What are your thoughts?

zou3519 replied:
Hmm, I like the proposal. Ideally I want some way to be able to roll the website back to a "last known good configuration" if we accidentally merge a bad change and are unable to fix it quickly.

One possible solution for that is: every week, on Sunday, have github actions do a manual commit to the gh-pages branch. The net effect of this is that we get roughly one new docs commit per week, and if something goes bad during the week, then we can roll the website back to its state from last week.

vfdev-5 replied:
Sounds good about cron-based GHA.

How about trying to make commit amend if previous commit was auto-generated and add a new commit otherwise ? In this case, we can safely roll back if needed and do not accumulated many auto-generated commits...
+ we can do everything cron-based (maybe daily).

zou3519 replied:
> How about trying to make commit amend if previous commit was auto-generated and add a new commit otherwise ? In this case, we can safely roll back if needed and do not accumulated many auto-generated commits... And we can do everything cron-based (maybe daily).

Okay! That sounds good. Hopefully we have not overengineered the system too much :)

vfdev-5 commented:
@zou3519 I updated the GHA and here is how it worked in practice on my fork:

- I merged this PR to main and the action made a new commit : https://github.com/vfdev-5/functorch/commit/6e5fc0bc34fb7b83ab006d0fbb1980940129eda7 "auto-generated commit"

<img width="1410" alt="Screen Shot 2022-03-15 at 18 22 25" src="https://user-images.githubusercontent.com/2459423/158440451-431bd0a0-787c-476b-b1d1-22261bc7f863.png">

- Next I updated a file in `docs` on `main` branch (https://github.com/vfdev-5/functorch/commit/ced0d43bd10f9ef314e216af442bd46f6fb868d3) and GHA pushed to `gh-pages` another auto generated commit with commit/amend - https://github.com/vfdev-5/functorch/commit/8f5fdb6d85c0554f1bf6fe3e1bca65feaad20e6d "auto-generated commit"

<img width="1417" alt="Screen Shot 2022-03-15 at 18 28 59" src="https://user-images.githubusercontent.com/2459423/158440630-ef26d7f5-011a-46bb-a81f-130420293550.png">

- Then I manually updated two files directly on gh-pages: https://github.com/vfdev-5/functorch/commit/792dfa29b8316388e319d5d90eac8297a3526d5a and https://github.com/vfdev-5/functorch/commit/16592dde7a695b7a78a253522a243b6ad5899702

<img width="1429" alt="Screen Shot 2022-03-15 at 18 36 58" src="https://user-images.githubusercontent.com/2459423/158441237-02d1003f-ef46-4ce5-beeb-94aa1fa3f734.png">

- Finally, I made another update on `main` : https://github.com/vfdev-5/functorch/commit/911447280772bd3e91ab38087d9a75fc46a96af2 and GHA pushed pushed to `gh-pages` another auto generated commit : https://github.com/vfdev-5/functorch/commit/8ead760319e3a8719fed48062158a7b7e4416473

<img width="1414" alt="Screen Shot 2022-03-15 at 18 48 37" src="https://user-images.githubusercontent.com/2459423/158441752-6979fb7c-6ebb-4482-90db-548c5084fea1.png">


So, I'd say things looks like working as expected.

zou3519 commented:
Thank you @vfdev-5! I might not get to reviewing this today, but will definitely do it by Monday

zou3519 reviewed (approved):
LGTM, fingers crossed

vfdev-5 commented:
Looks correct:
- https://github.com/pytorch/functorch/runs/5631772187?check_suite_focus=true
- https://github.com/pytorch/functorch/commit/7b7b5cfdfc4df8561470c24fabb4a47156ef9a96
- https://pytorch.org/functorch/nightly/

vfdev-5 commented:
@zou3519 docs deployment is failing now when try to overwrite auto-generated commit on gh-pages : https://github.com/pytorch/functorch/runs/5663088291?check_suite_focus=true

Looks like gh-pages is protected branch and "Cannot force-push to this protected branch"

zou3519 commented:
hmm, I thought I turned branch protection for the branch off. Let me check again
